src/main/java/io/orchestrate/client/Relation.java

```
@@ -1,33 +1,76 @@
+/*
+ * Copyright 2013 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package io.orchestrate.client;
 
 import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
 
-import lombok.AllArgsConstructor;
-import lombok.NonNull;
-import lombok.Getter;
+import lombok.EqualsAndHashCode;
+import lombok.ToString;
 
-@AllArgsConstructor
-public class Relation implements RelationMetadata {
+/**
+ * A container for a graph relation and its associated data.
+ *
+ * @param <T> The deserializable type for the value of the data belonging
+ *            to this event.
+ */
+@ToString(callSuper = true)
+@EqualsAndHashCode(callSuper = true)
+public class Relation<T> extends KvObject<T> implements RelationMetadata {
 
-    @NonNull @Getter
-    private final String sourceCollection;
-
-    @NonNull @Getter
-    private final String sourceKey;
-
-    @NonNull @Getter
     private final String destinationCollection;
-
-    @NonNull @Getter
     private final String destinationKey;
-
-    @NonNull @Getter
     private final String kind;
```

I suppose then we could consider introducing getKind() into the KvMetadata so that mixed-kind result lists can be checked with something other than instanceof. Though we might want to call it getKvKind? I don't really like that, but that would make it so that if someone was calling getKind on their relations before, they would have to change that to getRelation rather than it silently succeeding (but returning the path.kind instead of the relation).